File: examples/swift/ChessReferee/RefereeKit/Sources/RefereeKit/Referee.swift
import ChessKit
import Foundation

/// What the referee concluded from a stable board change.
public enum RefereeEvent: Equatable, Sendable {
    /// The game began: the starting position was recognized (in some rotation).
    case gameStarted

    /// A legal move was played. `fen` is the full FEN after the move.
    case legalMove(san: String, mover: String, fen: String)

    /// The observed position is not reachable by any legal move.
    case illegalMove(description: String, observedPlacement: String)

    /// A legal move — but by the side whose turn it isn't.
    case outOfTurn(san: String, mover: String)

    /// The board no longer matches; pieces changed in a way no single move
    /// explains (knocked pieces, mid-move hand, takeback).
    case boardScrambled(observedPlacement: String)
}

/// Turns a stream of noisy per-frame grids into confident game events.
///
/// Two jobs: a stability gate (a position must be read identically on
/// `stabilityThreshold` consecutive frames before it counts), and move
/// inference — diff the accepted position against the game state and find
/// the legal move that explains it, or rule the change illegal.
public final class Referee {
    public private(set) var board: Board
    private var orientation: Orientation = .unknown
    private var pending: PieceGrid?
    private var pendingCount = 0
    private let stabilityThreshold: Int

    public enum Orientation: Sendable {
        case unknown
        /// Camera row 0 is rank 8 (white at the bottom of the camera frame).
        case whiteBottom
        /// Camera row 0 is rank 1; the grid must be rotated 180°.
        case blackBottom
    }

    public init(stabilityThreshold: Int = 3) {
        self.stabilityThreshold = stabilityThreshold
        self.board = Board(position: .standard)
    }

    public var currentFEN: String { board.position.fen }
    public var sideToMove: String {
        board.position.sideToMove == .white ? "white" : "black"
    }

    /// Reset to a fresh game; orientation re-locks on the next recognized
    /// starting position.
    public func startNewGame() {
        board = Board(position: .standard)
        orientation = .unknown
        pending = nil
        pendingCount = 0
    }

    /// Feed one frame's grid. Returns an event only when a stable, *changed*
    /// position was accepted and classified.
    public func ingest(_ grid: PieceGrid) -> RefereeEvent? {
        // Stability gate: identical read N frames running.
        if grid == pending {
            pendingCount += 1
        } else {
            pending = grid
            pendingCount = 1
        }
        guard pendingCount == stabilityThreshold else { return nil }

        // Orientation locks by recognizing the starting position; until then
        // every stable grid is tested both ways.
        if orientation == .unknown {
            let startPlacement = placement(of: board.position.fen)
            if grid.fenPlacement == startPlacement {
                orientation = .whiteBottom
                return .gameStarted
            }
            if grid.rotated180.fenPlacement == startPlacement {
                orientation = .blackBottom
                return .gameStarted
            }
            return nil
        }

        let oriented = orientation == .whiteBottom ? grid : grid.rotated180
        let observed = oriented.fenPlacement
        guard observed != placement(of: board.position.fen) else { return nil }
        return classify(observed: observed, oriented: oriented)
    }

    // MARK: - Move inference

    private func classify(observed: String, oriented: PieceGrid) -> RefereeEvent {
        // Does any legal move for the side to move produce the observed
        // placement? Promotions are tried through every piece kind.
        if let explanation = legalMoveExplaining(observed, from: board) {
            let mover = board.position.sideToMove == .white ? "white" : "black"
            let san = apply(explanation, to: &board)
            return .legalMove(san: san, mover: mover, fen: board.position.fen)
        }

        // A legal move for the *other* side? That's a real board event worth
        // calling out differently: someone moved out of turn.
        var flipped = boardWithToggledSide(board)
        if let explanation = legalMoveExplaining(observed, from: flipped) {
            let mover = flipped.position.sideToMove == .white ? "white" : "black"
            let san = apply(explanation, to: &flipped)
            return .outOfTurn(san: san, mover: mover)
        }

        // No single legal move explains the change. A 1–4 square diff reads as
        // an attempted (illegal) move; anything bigger is a scrambled board.
        guard let current = PieceGrid(fenPlacement: placement(of: board.position.fen)) else {
            return .boardScrambled(observedPlacement: observed)
        }
        let diffs = oriented.changedSquares(from: current)
        guard (1...4).contains(diffs.count) else {
            return .boardScrambled(observedPlacement: observed)
        }
        return .illegalMove(
            description: describeIllegalChange(diffs, before: current, after: oriented),
            observedPlacement: observed
        )
    }

    private struct Explanation {
        let start: ChessKit.Square
        let end: ChessKit.Square
        let promotion: Piece.Kind?
    }

    /// Search every legal move (and promotion) of `base`'s side to move for
    /// one whose resulting placement matches `observed`.
    private func legalMoveExplaining(_ observed: String, from base: Board) -> Explanation? {
        let color = base.position.sideToMove
        for piece in base.position.pieces where piece.color == color {
            for target in base.legalMoves(forPieceAt: piece.square) {
                var candidate = base
                guard let move = candidate.move(pieceAt: piece.square, to: target) else {
                    continue
                }
                if isPromotion(move) {
                    for kind in [Piece.Kind.queen, .rook, .bishop, .knight] {
                        var attempt = candidate
                        _ = attempt.completePromotion(of: move, to: kind)
                        if placement(of: attempt.position.fen) == observed {
                            return Explanation(start: piece.square, end: target, promotion: kind)
                        }
                    }
                } else if placement(of: candidate.position.fen) == observed {
                    return Explanation(start: piece.square, end: target, promotion: nil)
                }
            }
        }
        return nil
    }

    /// Apply an explanation to `board`, returning the move's SAN.
    private func apply(_ explanation: Explanation, to board: inout Board) -> String {
        guard var made = board.move(pieceAt: explanation.start, to: explanation.end) else {
            return "\(explanation.start)-\(explanation.end)"
        }
        if let kind = explanation.promotion {
            made = board.completePromotion(of: made, to: kind)
        }
        return made.san
    }

    private func isPromotion(_ move: Move) -> Bool {
        move.piece.kind == .pawn && (move.end.rank.value == 1 || move.end.rank.value == 8)
    }

    // MARK: - Descriptions

    private func describeIllegalChange(
        _ diffs: [(row: Int, col: Int)], before: PieceGrid, after: PieceGrid
    ) -> String {
        var vacated: [(String, Character)] = []
        var arrived: [(String, Character)] = []
        for (row, col) in diffs {
            let square = squareName(row: row, col: col)
            let was = before.cells[row][col]
            let now = after.cells[row][col]
            if let was, now == nil { vacated.append((square, was)) }
            if let now, was == nil || was != now { arrived.append((square, now)) }
        }
        if let from = vacated.first, let to = arrived.first {
            return "\(pieceName(from.1)) from \(from.0) to \(to.0) — not a legal move here"
        }
        let squares = diffs.map { squareName(row: $0.row, col: $0.col) }.joined(separator: ", ")
        return "position changed on \(squares) in a way no legal move allows"
    }

    private func squareName(row: Int, col: Int) -> String {
        let file = Character(UnicodeScalar(UInt8(97 + col)))
        return "\(file)\(8 - row)"
    }

    private func pieceName(_ fenChar: Character) -> String {
        let color = fenChar.isUppercase ? "white" : "black"
        let names: [Character: String] = [
            "p": "pawn", "r": "rook", "n": "knight",
            "b": "bishop", "q": "queen", "k": "king",
        ]
        let kind = names[Character(fenChar.lowercased())] ?? "piece"
        return "\(color) \(kind)"
    }

    // MARK: - FEN helpers

    private func placement(of fen: String) -> String {
        String(fen.split(separator: " ").first ?? "")
    }

    private func boardWithToggledSide(_ base: Board) -> Board {
        var parts = base.position.fen.split(separator: " ").map(String.init)
        guard parts.count >= 2 else { return base }
        parts[1] = parts[1] == "w" ? "b" : "w"
        parts[3] = "-"
        guard let position = Position(fen: parts.joined(separator: " ")) else { return base }
        return Board(position: position)
    }
}

File: examples/typescript/chess-referee/src/referee.ts
import { Chess } from 'chess.js';

/** What the referee concluded from a stable board change. */
export type RefereeEvent =
  /** The game began: the starting position was recognized. */
  | { kind: 'game_started' }
  /** A legal move was played. `fen` is the full FEN after the move. */
  | { kind: 'legal_move'; san: string; mover: 'white' | 'black'; fen: string }
  /** The observed position is not reachable by any legal move. */
  | { kind: 'illegal_move'; description: string; observed: string }
  /** A legal move — but by the side whose turn it isn't. */
  | { kind: 'out_of_turn'; san: string; mover: 'white' | 'black' }
  /** Pieces changed in a way no single move explains (knocked pieces,
   *  mid-move hand, takeback). */
  | { kind: 'board_scrambled'; observed: string };

export const START_PLACEMENT = 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR';

export function placementOf(fen: string): string {
  return fen.split(' ')[0];
}

type Grid = (string | null)[][];

/** Expand a FEN placement into an 8×8 grid; row 0 is rank 8. */
function gridOf(placement: string): Grid | null {
  const ranks = placement.split('/');
  if (ranks.length !== 8) return null;
  const grid: Grid = [];
  for (const rank of ranks) {
    const row: (string | null)[] = [];
    for (const ch of rank) {
      if (ch >= '1' && ch <= '8') {
        for (let i = 0; i < Number(ch); i++) row.push(null);
      } else {
        row.push(ch);
      }
    }
    if (row.length !== 8) return null;
    grid.push(row);
  }
  return grid;
}

function squareName(row: number, col: number): string {
  return `${String.fromCharCode(97 + col)}${8 - row}`;
}

function pieceName(fenChar: string): string {
  const color = fenChar === fenChar.toUpperCase() ? 'white' : 'black';
  const names: Record<string, string> = {
    p: 'pawn', r: 'rook', n: 'knight', b: 'bishop', q: 'queen', k: 'king',
  };
  return `${color} ${names[fenChar.toLowerCase()] ?? 'piece'}`;
}

function changedSquares(before: Grid, after: Grid): { row: number; col: number }[] {
  const diffs: { row: number; col: number }[] = [];
  for (let row = 0; row < 8; row++) {
    for (let col = 0; col < 8; col++) {
      if (before[row][col] !== after[row][col]) diffs.push({ row, col });
    }
  }
  return diffs;
}

function describeIllegalChange(
  diffs: { row: number; col: number }[],
  before: Grid,
  after: Grid,
): string {
  const vacated: { square: string; piece: string }[] = [];
  const arrived: { square: string; piece: string }[] = [];
  for (const { row, col } of diffs) {
    const square = squareName(row, col);
    const was = before[row][col];
    const now = after[row][col];
    if (was !== null && now === null) vacated.push({ square, piece: was });
    if (now !== null && was !== now) arrived.push({ square, piece: now });
  }
  const from = vacated[0];
  const to = arrived[0];
  if (from && to) {
    return `${pieceName(from.piece)} from ${from.square} to ${to.square} — not a legal move here`;
  }
  const squares = diffs.map((d) => squareName(d.row, d.col)).join(', ');
  return `position changed on ${squares} in a way no legal move allows`;
}

/** Turns a stream of noisy board reads into confident game events.
 *
 *  Two jobs: a stability gate (a placement must be read identically on
 *  `stabilityThreshold` consecutive reads before it counts), and move
 *  inference — diff the accepted placement against the game state and find
 *  the legal move that explains it, or rule the change illegal.
 *
 *  Reads arrive already oriented (white's back rank last) — the vision
 *  endpoint resolves board orientation before we see the placement.
 */
export class Referee {
  private chess = new Chess();
  private started = false;
  private pending: string | null = null;
  private pendingCount = 0;

  constructor(private readonly stabilityThreshold: number = 2) {}

  get fen(): string {
    return this.chess.fen();
  }

  get placement(): string {
    return placementOf(this.chess.fen());
  }

  get sideToMove(): 'white' | 'black' {
    return this.chess.turn() === 'w' ? 'white' : 'black';
  }

  get gameStarted(): boolean {
    return this.started;
  }

  /** Reset to a fresh game; events resume once the starting position is
   *  recognized again. */
  startNewGame(): void {
    this.chess = new Chess();
    this.started = false;
    this.pending = null;
    this.pendingCount = 0;
  }

  /** Feed one read's placement. Returns an event only when a stable,
   *  *changed* position was accepted and classified. */
  ingest(placement: string): RefereeEvent | null {
    if (placement === this.pending) {
      this.pendingCount += 1;
    } else {
      this.pending = placement;
      this.pendingCount = 1;
    }
    if (this.pendingCount !== this.stabilityThreshold) return null;

    if (!this.started) {
      if (placement === START_PLACEMENT) {
        this.started = true;
        return { kind: 'game_started' };
      }
      return null;
    }

    if (placement === this.placement) return null;
    return this.classify(placement);
  }

  private classify(observed: string): RefereeEvent {
    const san = this.legalMoveExplaining(observed, this.chess.fen());
    if (san !== null) {
      const mover = this.sideToMove;
      this.chess.move(san);
      return { kind: 'legal_move', san, mover, fen: this.chess.fen() };
    }

    // A legal move for the *other* side? That's a real board event worth
    // calling out differently: someone moved out of turn. The game state
    // stays where it was — the move needs to be taken back.
    const toggled = this.toggledSideFen();
    if (toggled !== null) {
      const sanOther = this.legalMoveExplaining(observed, toggled);
      if (sanOther !== null) {
        const mover = this.sideToMove === 'white' ? 'black' : 'white';
        return { kind: 'out_of_turn', san: sanOther, mover };
      }
    }

    // No single legal move explains the change. A 1–4 square diff reads as
    // an attempted (illegal) move; anything bigger is a scrambled board.
    const before = gridOf(this.placement);
    const after = gridOf(observed);
    if (before === null || after === null) {
      return { kind: 'board_scrambled', observed };
    }
    const diffs = changedSquares(before, after);
    if (diffs.length < 1 || diffs.length > 4) {
      return { kind: 'board_scrambled', observed };
    }
    return {
      kind: 'illegal_move',
      description: describeIllegalChange(diffs, before, after),
      observed,
    };
  }

  /** Search every legal move of `fen`'s side to move for one whose resulting
   *  placement matches `observed`. Promotions are separate verbose moves, so
   *  every promotion piece is covered. */
  private legalMoveExplaining(observed: string, fen: string): string | null {
    const base = new Chess(fen);
    for (const move of base.moves({ verbose: true })) {
      const candidate = new Chess(fen);
      candidate.move(move.san);
      if (placementOf(candidate.fen()) === observed) return move.san;
    }
    return null;
  }

  /** Current position with the side to move flipped and en passant cleared,
   *  or null when the flip is not a valid position (e.g. the mover left the
   *  opponent in check). */
  private toggledSideFen(): string | null {
    const parts = this.chess.fen().split(' ');
    parts[1] = parts[1] === 'w' ? 'b' : 'w';
    parts[3] = '-';
    const fen = parts.join(' ');
    try {
      new Chess(fen);
      return fen;
    } catch {
      return null;
    }
  }
}
